### PR TITLE
Improve small sort performance on CUDA

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79627

Currently, `bitonicSortKVInPlace` is written to sort one array per
block of threads. If that dimension happens to be very small
(<128 elements), this results in low thread occupancy.

Instead, this changes `bitonicSortKVInPlace` to operate with a 2d
block. Sorting happens along the x dimension, and the y dimension
is a fixed size batch.

Running under `nvprof` shows speedups for both 32 and 128, which are the sizes changed here:

| Shape            | Master (ms) | This PR (ms) | Speedup |
|------------------|-------------|--------------|---------|
| (128, 1024, 32)  | 1.4794      | 0.9683       | 1.53    |
| (128, 1024, 128) | 6.8891      | 6.5528       | 1.05    |

Binary size is also marginally reduced, with my `torch_cuda.so` getting 500KB smaller.

### CODE REVIEW

**Comment by peterbell10** [comment]
Yes, there is absolutely a trade-off at small batch sizes.  Though a batch of 1024 is enough for this to be a significant win for 32 and a wash for 128.

| Shape            | Master (us) | This PR (us) | Speedup |
|------------------|-------------|--------------|---------|
| (128, 1024, 32)  | 1479.4      | 968.3        | 1.53    |
| (128, 1024, 128) | 6889.1      | 6552.8       | 1.05    |
| (1024, 32)       | 21.312      | 17.187       | 1.25    |
| (1024, 128)      | 79.680      | 79.608       | 1.00    |
| (128, 32)        | 7.7030      | 9.0610       | 0.85    |
| (128, 128)       | 15.041      | 18.735       | 0.80    |

File: aten/src/ATen/native/cuda/Sort.cu [hunk 1]
```diff
@@ -90,29 +89,29 @@ void sortKeyValueInplace(const TensorBase& key,
...
+        HANDLE_CASE(TYPE, A, 128, 4);                   \
+        break;                                          \
       case 32:                                          \
       case 16:                                          \
       case 8:                                           \
       case 4:                                           \
       case 2:                                           \
-      HANDLE_CASE(TYPE, A, 32);                         \
-      break;                                            \
+        HANDLE_CASE(TYPE, A, 32, 16);                   \
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/cuda/Sort.cu:108]: Oh actually, it should be possible to run the same kernel with a different runtime value for `block.y` as long as it's not greater than the templated `block_dim_y`. I'll see how that works out.

**Comment by peterbell10** [comment]
These new commits use the occupancy calculator functions to suggest the minimum grid size, then scales the batch size to achieve it. In my testing, this almost completely removed the slowdown on the small end.

| Shape            | Master (us) | This PR (us) | Speedup |
|------------------|-------------|--------------|---------|
| (128, 1024, 32)  | 1479.4      | 968.5        | 1.53    |
| (128, 1024, 128) | 6889.1      | 6553.3       | 1.05    |
| (1024, 32)       | 21.312      | 15.584       | 1.37    |
| (1024, 128)      | 79.680      | 79.136       | 1.01    |
| (128, 32)        | 7.7030      | 7.744        | 0.99    |
| (128, 128)       | 15.041      | 14.976       | 1.00    |

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**